src/openloco/windows/construction/Common.cpp

```diff
@@ -0,0 +1,2113 @@
+                window->widgets[construction::widx::construct].type = widget_type::wt_3;
+                window->widgets[construction::widx::construct].tooltip = string_ids::tooltip_construct;
+                window->widgets[construction::widx::rotate_90].type = widget_type::wt_9;
+                window->widgets[construction::widx::rotate_90].image = image_ids::construction_new_position;
+                window->widgets[construction::widx::rotate_90].tooltip = string_ids::new_construction_position;
+            }
+            if (_constructionHover == 0 || _constructionHover == 1)
+            {
+                if (_lastSelectedTrackPiece != 0xFF)
+                {
+                    auto trackPieceWidget = trackPieceWidgets[_lastSelectedTrackPiece];
+                    activatedWidgets |= 1ULL << trackPieceWidget;
+                }
+
+                uint8_t trackGradient = construction::widx::level;
+
+                switch (_lastSelectedTrackGradient)
+                {
+                    case common::trackGradient::level:
+                        trackGradient = construction::widx::level;
+                        break;
+
+                    case common::trackGradient::slope_up:
+                        trackGradient = construction::widx::slope_up;
+                        break;
+
+                    case common::trackGradient::slope_down:
+                        trackGradient = construction::widx::slope_down;
+                        break;
+
+                    case common::trackGradient::steep_slope_up:
+                        trackGradient = construction::widx::steep_slope_up;
+                        break;
+
+                    case common::trackGradient::steep_slope_down:
+                        trackGradient = construction::widx::steep_slope_down;
+                        break;
+                }
+
+                activatedWidgets |= 1ULL << trackGradient;
+            }
+            window->activated_widgets = activatedWidgets;
+            window->invalidate();
+        }
+
+        static void activateSelectedTrackWidgets(window* window)
+        {
+            tilemgr::map_invalidate_map_selection_tiles();
+            _mapSelectionFlags = _mapSelectionFlags | (1 << 3) | (1 << 1);
+
+            auto track = getTrackPieceId(_lastSelectedTrackPiece, _lastSelectedTrackGradient, _constructionRotation);
+
+            uint8_t rotation;
+            uint8_t trackId;
+            auto x = _x;
+            auto y = _y;
+
+            if (!track)
+            {
+                rotation = _constructionRotation;
+                trackId = 0;
+            }
+            else
+            {
+                rotation = track->rotation;
+                trackId = track->id;
+            }
+
+            auto trackPiece = trackPieces[trackId];
+            auto i = 0;
+            auto posId = 0;
+
+            while (trackPiece[i].index != 0xFF)
+            {
+                if (trackPiece[i].flags & (1 << 7))
```

Comment:
If track is diagonal then you don't have the extra tile infront. I came across a very similar design in OpenRCT2. So you can check this with diagonal track.